projects/vdk-core/src/vdk/api/plugin/plugin_input.py

```diff
@@ -165,9 +165,72 @@ def initialize_job(context: JobContext) -> None:
 
 
 class IIngesterPlugin:
+    """
+    Interface representing the APIs that can be used when creating Ingester
+    Plugins. Plugins implement one of the methods, depending on their needs.
+    If more complex processing is required, plugins can be chained
+    together.
+
+    The lifecycles of the three methods are as follows:
+
+    ingest_payload() - plugins implementing this method are intended
+                       to do actual data ingestion. Versatile Data Kit will
+                       call this method when the payload that data engineers
+                       have passed is ready to be ingested.
+
+    process_payload() - plugins implementing this method do not aim at
+                        ingesting the payload, rather to do some processing
+                        of the payload before it is ingested.
+
+    post_ingest_process() - plugins implementing this method aim at doing
+                            some metadata processing after the payload is
+                            ingested. The main use-case is telemetry data
+                            collection.
+                            NOTE: This method can be used only by plugins
+                            that implement ingest_payload().
+
+    Example plugin chaining workflows:
+
+    Workflow 1
+    ----------
+    _____________________      _________________________      ______________
+    |      Data Job     |      |    vdk-ingest-http    |      |            |
+    |-------------------|      |-----------------------|      | Data Lake/ |
+    |                   |      |                       |      | Warehouse/ |
+    |send_object_       | ---> |   ingest_payload()    | ---> | Third-party|
```

Comment:
It seems to me pydocs would rather benefit from a (pseudo)code snippet in favour of ascii art.
Maybe we could also add a page in the wiki about ingestion, pre- and post-processing.
I would suggest adding to PR test coverage that ensures the newly added methods are evaluated when implementation provided. That would be the full pre-process->ingest->post-process usage example (happy path), that also verifies chaining.

Reply:
@ivakoleva Could you elaborate what examples you're looking for? Generally, method definitions contain code examples: https://github.com/vmware/versatile-data-kit/pull/579/files#diff-65fe93ce6e4724edf9ad045d4ac6c1bc46a189cabf33ab6cc9eb4a00304d68aeR289

Reply:
I like the ASCII art. I think it clearly shows the data flow.

Reply:
But I agree a bit more about code snippet. But perhaps we can make an example for this - https://github.com/vmware/versatile-data-kit/wiki/Examples

The thing that is not clear   is what would be the implementation of vdk-ingest-http. 
For example vdk-ingest-http implements only ingest_payload ? While vdk-validate would implement only process_payload?

Reply:
@tozka With what we have now, **vdk-validate** would need to implement only process_payload(), while **vdk-ingest-http** would need to implement *ingest_payload()* and *post_ingest_process()* ([ingester base implementation](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/builtin_plugins/ingestion/ingester_base.py#L526))

If we want to have complete separation, i.e., different plugins to implement only one method (e.g., **vdk-validate** -> *process_payload()*; **vdk-ingest-http** -> *ingest_payload()*; **vdk-telemetry** -> *post_ingest_process()*), then we would need to do substantial redesign of **IngesterRouter** and **IngesterBase**.

Reply:
We need vdk-telemetry to be separated as we'd need same telemetry regardless of database used (we know of use-cases where it would be necessary for both impala and trino). 

But I see no need of any substantial redesign . 

The default implementation of all methods could be -a no-op . For example: 
```
def ingest_payload(): 
      pass 
def process_payload(payload)
      # no -op return same payload
      return payload
 ```

Then a plugin can choose which method to overwrite

Reply:
I agree that having separate plugins is the best option. My concern, however, is that if for example, the **vdk-ingest-http** plugin wants to calculate the size of the payload before the post request is made, and pass it to the telemetry plugin, we would need to add functionality to return that information to the IngestRouter, and from there, pass it to the telemetry plugin. This may have performance penalty which we would need to account for.

Reply:
This should not be vehichle for communication between plugins. Plugin need to be independant. If telemetry plugin needs to know the size of the payload it can calculate it itself.

If trino plugin wants to send trino specific telemetry it'd need to hook itself to telemetry hooks provided by telemetry plugin perhaps. But that's out of scope for this change.

Reply:
Added tests as discussed offline.